### PR TITLE
Process improvements

### PR DESCRIPTION
**Pull request by jhass**

### CODE REVIEW

File: src/process.cr [hunk 1]
```diff
@@ -47,8 +47,12 @@ class Process
     nil
   end
 
-  def self.ppid : LibC::PidT
```
Comment (RX14) [inline src/process.cr]: Why is this removed?

Comment (RX14) [inline src/process.cr]: Ah, looks like you moved the `ppid` method in the file and the addition and deletion got split between two commits.

Comment (jhass) [inline src/process.cr]: Right, made a mistake while committing, fixed.

File: src/process.cr [hunk 2]
```diff
@@ -5,24 +5,40 @@ require "c/sys/wait"
...
     LibC.getpid
   end
 
+  # Returns the process group identifier of the current process.
   def self.pgid : LibC::PidT
     pgid(0)
   end
 
+  # Returns the process group identifier of the process with the process
+  # identifier *pid*.
```
Comment (ysbaddaden) [inline src/process.cr]: What about:

> Returns the process group identifier of another process identified by `pid`?

File: spec/std/process_spec.cr [hunk 3]
```diff
@@ -169,4 +170,25 @@ describe Process do
...
+    fork.wait
+
+    File.exists?(tmpfile.path).should be_true
+    tmpfile.unlink
+  end
+
+  it "checks for existence" do
+    Process.exists?(Process.ppid).should be_true
+    # We can't reliably check whether it ever returns false, since we can't predict
+    # how PIDs are used by the system.
```
Comment (ysbaddaden) [inline spec/std/process_spec.cr]: What about:

```crystal
child = Process.fork { sleep 1 }
Process.exists?(child.pid).should be_true
child.kill
child.wait
Process.exists?(child.pid).should be_false
```

BTW: a `child.exists?` method would be nice :)

File: src/process.cr [hunk 4]
```diff
@@ -31,19 +47,26 @@ class Process
     nil
   end
 
-  def self.ppid : LibC::PidT
-    LibC.getppid
+  # Returns true if the process identifier given by *pid* is valid for
+  # a currently running process, false otherwise.
+  def self.exists?(pid : Int)
```
Comment (ysbaddaden) [inline src/process.cr]: From the linux man page (conforming to POSIX), if the process or process group does not exist then errno is set to `ESRCH`, so shouldn't this be:

```crystal
def self.exists?(pid : Int)
  if LibC.kill(pid, 0) == 0
    return true
  end
  if Errno.value == Errno::ESRCH
    return false
  end
  raise Errno.new("kill")
end
```

**Comment by ysbaddaden** [comment]
What about moving the Process instance methods from `src/process/run.cr` to `src/process.cr` while we're at it?

**Comment by ysbaddaden** [comment]
Nice refactor, just a few improvements and it's :+1: to merge from me!

File: spec/std/process_spec.cr (hunk 3 above)
Comment (jhass) [inline spec/std/process_spec.cr]: The pid becomes free for reuse immediately, we have no guarantee that a new process is spawned for something at the same time reusing it.

**Comment by jhass** [comment]
If anything I would collapse all of `run.cr` into `process.cr`, the instance methods make more sense in `run.cr` to me since it also contains the primary way of constructing a `Process` instance in the first place, while `process.cr` currently just has class level helpers. I'm on the fence of whether I'd like them combined, so should I do it here or not?

**Comment by asterite** [comment]
I wouldn't mind merging the two files into one

File: spec/std/process_spec.cr (hunk 3 above)
Comment (ysbaddaden) [inline spec/std/process_spec.cr]: On Linux PID appear to increment and eventually overflow, which would avoid this issue of PID reuse, especially on a fast race? I don't know about other systems, but I hope they don't reuse a PID immediately after freeing...

File: src/process.cr [hunk 5]
```diff
@@ -85,19 +128,351 @@ class Process
...
+  private def channel
+    @channel ||= Channel(Exception?).new
+  end
+
+  private def needs_pipe?(io)
+    io.nil? || (io.is_a?(IO) && !io.is_a?(IO::FileDescriptor))
+  end
+
+  private def exec_io_for(io)
+    # TODO: Compiler bug: The union doesn't properly degrade to IO::FileDescriptior|Bool
```
Comment (asterite) [inline src/process.cr]: @jhass why is this method needed? If I replace this method body with just `io` it continues to compile and run just fine.

File: src/process.cr [hunk 6]
```diff
@@ -85,19 +128,351 @@ class Process
...
+          env,
+          clear_env,
+          exec_io_for(fork_input || input),
+          exec_io_for(fork_output || output),
+          exec_io_for(fork_error || error),
+          chdir
+        )
+      rescue ex
+        ex.inspect_with_backtrace STDERR
+        LibC._exit 127 # TODO: Compiler bug: ensure isn't run when an exception is raised and catched here
```
Comment (asterite) [inline src/process.cr]: @jhass Any way this can be reduced and reported as a separate bug?

File: src/process.cr (hunk 5 above)
Comment (jhass) [inline src/process.cr]: Did you try `make spec`?

File: src/process.cr (hunk 6 above)
Comment (jhass) [inline src/process.cr]: I didn't try yet and not sure I find the motivation tbh.

Comment (asterite) [inline src/process.cr]: Here: https://github.com/crystal-lang/crystal/issues/3082

File: src/process.cr (hunk 5 above)
Comment (jhass) [inline src/process.cr]: So I can remove the method and leave out the restrictions, working around the bug in another way, or add the restrictions so the method becomes necessary again and serves as a reminder about the bug. What do you prefer?

Comment (asterite) [inline src/process.cr]: Thanks, I was able to reproduce it, I'll try to see what's going on. Later I'll report an issue, for now you can leave it as you prefer.

**Comment by asterite** [comment]
Let's drop `exec` and just have `run` and a `replace` flag.

Our original `Process.run` design had just one `run` method. In Ruby there are tons of ways to do the same thing and it's pretty confusing, that's why we aimed for just one method with many options.

File: src/process.cr (hunk 5 above)
Comment (asterite) [inline src/process.cr]: @jhass Mmm... `input` can be nil, the restriction on `run` is `Stdio` which includes `Nil`. So I think the error is correct. Or am I missing something?

Comment (jhass) [inline src/process.cr]: If it's indeed `nil`, `needs_pipe?` returns true, so `fork_input` is taken, I guess the actual bug was that a `not_nil!` didn't help either.

**Comment by jhass** [comment]
`Process.run` and `Process.exec` are not the same thing and they can't share the same interface, in particular passing `input: nil` etc. to `exec` makes no sense whatsoever.

**Comment by asterite** [comment]
Can we then name it `Process.replace`? `run` and `exec` have basically the same meaning in English...

**Comment by jhass** [comment]
`exec` is well known in UNIX to replace the current process, the system call is `execve` (`execvp` etc are wrappers around it). Ruby uses `exec` for this operation. In Python it's `os.execvpe` and friends, `pcntl_exec` in PHP, `kexec` in node via the npm package of the same name, basically all standard libraries that provide an interface to do it, have `exec` in its name.

File: src/process.cr (hunk 5 above)
Comment (jhass) [inline src/process.cr]: Well, it brings something back, I can no longer reproduce it complaining about `nil` instead of `MemoryIO`:

```diff
diff --git a/src/process.cr b/src/process.cr
index 5d9bff5..c683032 100644
--- a/src/process.cr
+++ b/src/process.cr
@@ -235,9 +235,9 @@ class Process
           argv,
           env,
           clear_env,
-          fork_input || input,
-          fork_output || output,
-          fork_error || error,
+          (fork_input || input).not_nil!,
+          (fork_output || output).not_nil!,
+          (fork_error || error).not_nil!,
           chdir
         )
       rescue ex
@@ -360,7 +360,7 @@ class Process
   end
 
   # :nodoc:
-  protected def self.exec_internal(command : String, argv, env, clear_env, input, output, error, chdir)
+  protected def self.exec_internal(command : String, argv, env, clear_env, input : Bool|IO::FileDescriptor, output : Bool|IO::FileDescriptor, error  : Bool|IO::FileDescriptor, chdir)
     reopen_io(input, STDIN, "r")
     reopen_io(output, STDOUT, "w")
     reopen_io(error, STDERR, "w")
```
`make spec` errors with:

```
in ./src/process.cr:233: no overload matches 'Process.exec_internal' with types String, Array(Pointer(UInt8)), Nil, Bool, (IO::FileDescriptor | MemoryIO), IO::FileDescriptor, IO::FileDescriptor, Nil
Overloads are:
 - Process.exec_internal(command : String, argv, env, clear_env, input : Bool | IO::FileDescriptor, output : Bool | IO::FileDescriptor, error : Bool | IO::FileDescriptor, chdir)
Couldn't find overloads for these types:
 - Process.exec_internal(command : String, argv : Array(Pointer(UInt8)), env : Nil, clear_env : Bool, input : MemoryIO, output : IO::FileDescriptor, error : IO::FileDescriptor, chdir : Nil)

        Process.exec_internal(
                ^~~~~~~~~~~~~
```

Since casting to a sub-union isn't possible, this was my solution to tell Crystal what the actually possible types are at this point.

```diff
diff --git a/src/process.cr b/src/process.cr
index 5d9bff5..e56defa 100644
--- a/src/process.cr
+++ b/src/process.cr
@@ -235,9 +235,9 @@ class Process
           argv,
           env,
           clear_env,
-          fork_input || input,
-          fork_output || output,
-          fork_error || error,
+          (fork_input || input).as(Bool|IO::FileDescriptor),
+          (fork_output || output).as(Bool|IO::FileDescriptor),
+          (fork_error || error).as(Bool|IO::FileDescriptor),
           chdir
         )
       rescue ex
@@ -360,7 +360,7 @@ class Process
   end
 
   # :nodoc:
-  protected def self.exec_internal(command : String, argv, env, clear_env, input, output, error, chdir)
+  protected def self.exec_internal(command : String, argv, env, clear_env, input : Bool|IO::FileDescriptor, output : Bool|IO::FileDescriptor, error  : Bool|IO::FileDescriptor, chdir)
     reopen_io(input, STDIN, "r")
     reopen_io(output, STDOUT, "w")
     reopen_io(error, STDERR, "w")
```
actually crashes the compiler now that I try it.

**Comment by asterite** [comment]
I don't think everyone is familiar with UNIX, and having `exec` and `run` will only bring confusion. What's wrong with `Process.replace`? The docs say "Replaces the current process", so I see no reason to stick with `exec`. That's also the reason why we don't use `recv` in `Socket`. I prefer to have meaningful names over anecdotical, meaningless names.

File: src/process.cr [hunk 7]
```diff
@@ -5,45 +5,87 @@ require "c/sys/wait"
 require "c/unistd"
 
 class Process
+  # Terminate the current process immediately. All open files, pipes and sockets
+  # are flushed and closed, all child processes are inherited by PID 1. This does
+  # not run any handlers registered with `at_exit`, use `::exit` for that.
+  #
+  # *status* is the exit status of the current process.
   def self.exit(status = 0)
```
Comment (ysbaddaden) [inline src/process.cr]: Aparte:

- what about renaming it to `Process.exit!` like in Ruby, to mark the difference and avoid a confusion?
- `::exit(bool)` would be nice sugar too.

If those are interesting I'll propose a follow-up PR.

**Comment by jhass** [comment]
This is about familiarity, I brought examples of where `exec` is used, do you have examples of where `replace` or something else is used? The point is that if you're going to use this operation, you're most likely already know it from somewhere else and you'll know it as `exec` and look for that first.

**Comment by ysbaddaden** [comment]
`Process.replace` makes me think that I can replace any process by another process, which feels weird, whereas `Process.exec` feels like I'm executing something, thought one that isn't familiar with UNIX may stumble on the fact that it replaces the current process, but it's nothing that documentation can't fix. I admit I may be conservative here, but this is a POSIX only feature that only POSIX developers are aware of, so... this feels like renaming `fork` as `duplicate`: weird o.O

**Comment by asterite** [comment]
Let's leave it as `exec` for now, I guess with so many ways to spawn processes there will always be a bit of confusion (for example we have `Process.run`, `backtick` and `system`).

**Comment by jhass** [comment]
Then this is done from my side :)

**Comment by drosehn** [comment]
Maybe `Process.replace_exec` or `Process.replace_me`?  😃

**Comment by asterite** [comment]
@jhass Thank you! And thank you for also cleaning up the code and documenting it. It's super readable now!
